README
NAME
    Net::Works - Sane APIs for IP addresses and networks

VERSION
    version 0.19

DESCRIPTION
    The NetAddr::IP module is very complete, correct, and useful. However,
    its API design is a bit crufty. This distro provides an alternative API
    that aims to address the biggest problems with that module's API, as
    well as adding some additional features.

    This distro contains two modules, Net::Works::Address and
    Net::Works::Network.

    NOTE: This distro's APIs are still in flux. Use at your own risk.

Net::Works VERSUS NetAddr::IP
    Here are some of the key differences between the two distributions:

    *   Separation of address from network

        "Net::Works" provides two classes, one for single IP addresses and
        one for networks (and subnets). With NetAddr::IP a single address is
        represented as a /32 or /128 subnet. This is technically correct but
        can make the API harder to use. Whenever you want a single IP you're
        always stuck checking that the object you're working with is the
        size of a single address subnet.

    *   Multiple constructors

        Net::Works allows you to construct an IP address from a string
        ("192.0.2.1") or an integer (3221225985) using separate
        constructors.

    *   Next & previous IP

        You can get the next and previous address from Net::Works::Address
        object, regardless of whether or not that address is in the same
        subnet.

    *   Constructors throw exceptions

        If you pass bad data to a constructor you'll get an exception.

    *   Sane iterator and first/last

        The iterator provided by Net::Works::Network has no confusing
        special cases. It always returns all the addresses in a network,
        including the network and broadcast addresses. Similarly, the
        "$network->first()" and "$network->last()" do not return different
        results for different sized networks.

    *   Split a range into subnets

        The Net::Works::Network class provides a
        "Net::Works::Network->range_as_subnets" method that takes a start
        and end IP address and splits this into a set of subnets that
        include all addresses in the range.

    *   Does less

        This distro does not support every method provided by NetAddr::IP.
        Patches to add more features are welcome, however.

BUGS
    Please report any bugs or feature requests to
    "bug-net-works@rt.cpan.org", or through the web interface at
    <http://rt.cpan.org>. I will be notified, and then you'll automatically
    be notified of progress on your bug as I make changes.

AUTHORS
    *   Dave Rolsky <autarch@urth.org>

    *   Greg Oschwald <oschwald@cpan.org>

    *   Olaf Alders <oalders@wundercounter.com>

CONTRIBUTORS
    *   Alexander Hartmaier <abraxxa@cpan.org>

    *   Gregory Oschwald <oschwald@gmail.com>

    *   TJ Mather <tjmather@maxmind.com>

    *   William Stevenson <wstevenson@maxmind.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by MaxMind, Inc..

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
{{$NEXT}}

- namespace::autoclean is now used to clean module namespaces.

- confessions now correctly skip internal stack frames

- Subclasses of Net::Works::Addresses correctly return instances of the subclass
  when next_ip and previous_ip are called, and as_ipv4_string correctly respects
  overridden new_from_integer methods.  Thanks to edmundadjei for pointing out
  this issue and helping develop a fix. GH #23.


0.22     2016-08-21

- Added a Net::Works::Network->is_single_address method. This returns true if
  the network contains just one single address. Requested byAlexander
  Hartmaier. GH #17.


0.21     2015-04-30

- `new_from_integer` will now accept a `Math::BigInt` in addition to a
  `Math::UInt128` or a standard Perl integer.


0.20     2014-10-16

- Added numeric & string comparison overloading to Net::Works::Network.

- Added string comparison overloading to Net::Works::Address.


0.19     2014-08-27

- Fixed validation when a string with more than one slash (/) is passed to
  Net::Works::Network->new_from_string.


0.18     2014-07-08

- Fixed warnings that would happen when trying to validate some bad data. This
  could happen if you passed a string with no slash (/) to
  Net::Works::Network->new_from_string.

- Make all errors call Carp::confess instead of die for consistency.


0.17     2014-07-03

- The mask_length attribute (and related methods) in Net::Works::Network has
  been renamed to prefix_length, which is more correct and consistent with the
  terminology used for IPv6. The old names (and methods) are still available,
  but will probably start warning in a future release. Fixes GH #8.

- Require namespace::autoclean 0.16+ so our classes are not inflated from Moo
  to Moose classes. Fixes GH #6.


0.16     2013-12-25

- The last address in the ::0/0 network was being returned as ::0 instead of
  ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff.

- The IPv6 address ::0 is now always stringified as "::0", not "::". However,
  either will be parsed correctly when giving an address as a string to a
  constructor.


0.15     2013-12-23

- If you passed a Math::Int128 number to Net::Works::Network->new_from_integer
  and specified the ip version as 4, the object would still stringify as IPv6.


0.14     2013-12-17

- Docs were changed to only use IPv4 and IPv6 addresses from the documentation
  address space in the docs. (Alexander Hartmaier)

- Net::Works::Network objects now overload stringification. (Alexander
  Hartmaier)

- Added more reserved IPv4 and IPv6 networks to be removed from a given IP
  range and included a list of all such networks in the docs.


0.13     2013-08-22

- Added two new Net::Works::Network methods: $network->contains() and
  $network->split().


0.12     2013-07-10

- Converted to use Moo instead of Moose.

- Switched bugtracker to GitHub.


0.11     2013-06-24

- Added $network->first_as_integer() and ->last_as_integer() methods. This is
  an optimization so you don't have to write
  $network->first()->as_integer(). Internally, the latter just turns an
  integer into a Net::Works::Address object that you then get the same integer
  out of.


0.10     2013-06-20

- The Net::Works::Network->range_as_subnets() method now takes a version
  parameter so you can explicitly ask for IPv4 or IPv6.

- Also fixed the license in the docs to match the LICENSE file in the repo
  (same as Perl 5).


0.09     2013-02-06

- Documentation fixes


0.08     2013-02-04

- Fixed type constraints so they inline.


0.07     2012-12-11

- Removed all use of Data::Validate::IP and NetAddr::IP::Util.

- Validate all data passed to the constructor.


0.06     2012-12-07

- Require Socket 1.99+. This fixes test failures with old version of Socket
  that don't provide inet_ntop or inet_pton.


0.05     2012-12-07

- More internals rewrites for the sake of speed. We now use Math::Int128
  instead of Math::BigInt. The former is way faster.

- Net::Works::Network now has two constructors, new_from_string() and
  new_from_integer(), just like Net::Works::Address. The new() method should
  no longer be used.


0.04     2012-12-04

- Speed improvements for Net::Works::Network->range_as_subnets(). For most use
  cases this won't matter but for our use at MaxMind we end up calling this a
  lot so making it faster is a big win.


0.03     2012-12-04

- Work around a bug in Data::Validate::IP's is_ipv6 sub. It doesn't think '::'
  is a valid IPv6 address. If you created a Net::Works::Network object with
  this address it would treat it as an IPv4 address unless you explicitly
  specified the version.


0.02     2012-12-03

- All changes in this release come from Greg Oschwald.

- Major internals rewrite to speed things up. We no longer wrap NetAddr::IP
  objects (but we do use NetAddr::IP::Util).

- Net::Works::Network->mask_length() now returns the mask as passed to the
  constructor, not the number of bits in the mask. The ->bits method returns
  the number of bits in the mask (32 or 128).

- The Network->netmask_as_integer() method no longer exists (use mask_length()
  instead). The ->max_netmask_as_integer() method has been renamed to
  ->max_mask_length().

- Added Net::Works::Address->mask_length() and ->bits() (which always return
  the same thing).

- We use Socket::inet_ntop to stringify addresses.


0.01     2012-11-28

- First release upon an unsuspecting world.


lib/Net/Works.pm
package Net::Works;

use strict;
use warnings;
use 5.010000;

our $VERSION = '0.23';

1;

# ABSTRACT: Sane APIs for IP addresses and networks

__END__

=head1 DESCRIPTION

The L<NetAddr::IP> module is very complete, correct, and useful. However, its
API design is a bit crufty. This distro provides an alternative API that aims
to address the biggest problems with that module's API, as well as adding some
additional features.

This distro contains two modules, L<Net::Works::Address> and
L<Net::Works::Network>.

B<NOTE: This distro's APIs are still in flux. Use at your own risk.>

=head1 Net::Works VERSUS NetAddr::IP

Here are some of the key differences between the two distributions:

=over 4

=item * Separation of address from network

C<Net::Works> provides two classes, one for single IP addresses and one for
networks (and subnets). With L<NetAddr::IP> a single address is represented as
a /32 or /128 subnet. This is technically correct but can make the API harder
to use. Whenever you want a single IP you're always stuck checking that the
object you're working with is the size of a single address subnet.

=item * Multiple constructors

L<Net::Works> allows you to construct an IP address from a string ("192.0.2.1")
or an integer (3221225985) using separate constructors.

=item * Next & previous IP

You can get the next and previous address from L<Net::Works::Address> object,
regardless of whether or not that address is in the same subnet.

=item * Constructors throw exceptions

If you pass bad data to a constructor you'll get an exception.

=item * Sane iterator and first/last

The iterator provided by L<Net::Works::Network> has no confusing special
cases. It always returns all the addresses in a network, including the network
and broadcast addresses. Similarly, the C<< $network->first() >> and C<<
$network->last() >> do not return different results for different sized networks.

=item * Split a range into subnets

The L<Net::Works::Network> class provides a C<<
Net::Works::Network->range_as_subnets >> method that takes a start and end IP
address and splits this into a set of subnets that include all addresses in
the range.

=item * Does less

This distro does not support every method provided by L<NetAddr::IP>. Patches
to add more features are welcome, however.

=back

=head1 BUGS

Please report any bugs or feature requests to C<bug-net-works@rt.cpan.org>, or
through the web interface at L<http://rt.cpan.org>.  I will be notified, and
then you'll automatically be notified of progress on your bug as I make
changes.


lib/Net/Works/Util.pm
package Net::Works::Util;

use strict;
use warnings;

our $VERSION = '0.23';

use Carp qw( confess );
use Math::Int128 qw( net_to_uint128 uint128_to_net );
use Socket qw( AF_INET AF_INET6 inet_pton inet_ntop );
use Scalar::Util qw( blessed );

use namespace::clean;

use Exporter qw( import );

our @EXPORT_OK = qw(
    _string_address_to_integer
    _integer_address_to_binary
    _binary_address_to_string
    _integer_address_to_string
    _validate_ip_string
);

## no critic (Subroutines::ProhibitUnusedPrivateSubroutines)
sub _string_address_to_integer {
    my $string  = shift;
    my $version = shift;

    my $binary = inet_pton( $version == 4 ? AF_INET : AF_INET6, $string )
        or return;

    return $version == 4
        ? unpack( N => $binary )
        : net_to_uint128($binary);
}

sub _integer_address_to_binary {
    my $integer = shift;

    if ( ref $integer && blessed $integer) {
        return uint128_to_net($integer);
    }
    else {
        return pack( N => $integer );
    }
}

sub _binary_address_to_string {
    my $binary = shift;

    my $family = length($binary) == 4 ? AF_INET : AF_INET6;

    my $string = inet_ntop( $family, $binary );
    return $string eq '::' ? '::0' : $string;
}

sub _integer_address_to_string {
    _binary_address_to_string( _integer_address_to_binary( $_[0] ) );
}

sub _validate_ip_string {
    my $str     = shift;
    my $version = shift;

    my $str_val = defined $str ? $str : 'undef';
    if ( $version == 4 ) {
        confess("$str_val is not a valid IPv4 address")
            unless defined $str && defined inet_pton( AF_INET, $str );
    }
    else {
        confess("$str_val is not a valid IPv6 address")
            unless defined $str && defined inet_pton( AF_INET6, $str );
    }
}

1;

# ABSTRACT: Utility subroutines for Net-Works

__END__

=head1 DESCRIPTION

All of the subroutines in this module are really just for our internal use. No
peeking.

=cut


lib/Net/Works/Types.pm
package Net::Works::Types;

use strict;
use warnings;

our $VERSION = '0.23';

use Carp qw( confess );
use Scalar::Util ();
use Sub::Quote qw( quote_sub );

use namespace::clean;

use Exporter qw( import );

our @EXPORT_OK = qw(
    Int
    IPInt
    IPVersion
    PrefixLength
    NetWorksAddress
    PackedBinary
    Str
);

## no critic (ValuesAndExpressions::ProhibitImplicitNewlines, NamingConventions::Capitalization)
{
    my $t = quote_sub(
        q{
( defined $_[0] && !ref $_[0] && $_[0] =~ /^[0-9]+\z/ )
    or Net::Works::Types::_confess(
    '%s is not a valid integer for an IP address',
    $_[0]
    );
}
    );

    sub Int () {$t}
}

{
    my $t = quote_sub(
        q{
(
    defined $_[0] && ( ( !ref $_[0] && $_[0] =~ /^[0-9]+\z/ )
        || ( Scalar::Util::blessed( $_[0] )
            && ( $_[0]->isa('Math::UInt128') || $_[0]->isa('Math::BigInt') ) )
        )
    )
    or Net::Works::Types::_confess(
    '%s is not a valid integer for an IP address',
    defined $_[0] ? $_[0] : 'undef'
    );
}
    );

    sub IPInt () {$t}
}

{
    my $t = quote_sub(
        q{
( defined $_[0] && !ref $_[0] && ( $_[0] == 4 || $_[0] == 6 ) )
    or Net::Works::Types::_confess(
    '%s is not a valid IP version (4 or 6)',
    defined $_[0] ? $_[0] : 'undef'
    );
        }
    );

    sub IPVersion () {$t}
}

{
    my $t = quote_sub(
        q{
( !ref $_[0] && defined $_[0] && $_[0] =~ /^[0-9]+\z/ && $_[0] <= 128 )
    or Net::Works::Types::_confess(
    '%s is not a valid IP network prefix length (0-128)',
    defined $_[0] ? $_[0] : 'undef'
    );
}
    );

    sub PrefixLength () {$t}
}

{
    my $t = quote_sub(
        q{
( Scalar::Util::blessed( $_[0] ) && $_[0]->isa('Net::Works::Address') )
    or Net::Works::Types::_confess(
    '%s is not a Net::Works::Address object',
    $_[0]
    );
}
    );

    sub NetWorksAddress () {$t}
}

{
    my $t = quote_sub(
        q{
( defined $_[0] && !ref $_[0] )
    or Net::Works::Types::_confess( '%s is not binary data', $_[0] );
}
    );

    sub PackedBinary () {$t}
}

{
    my $t = quote_sub(
        q{
( defined $_[0] && !ref $_[0] )
    or Net::Works::Types::_confess( '%s is not a string', $_[0] );
}
    );

    sub Str () {$t}
}
## use critic

## no critic (Subroutines::ProhibitUnusedPrivateSubroutines, Variables::ProhibitPackageVars)
sub _confess {
    local $Carp::Internal{ (__PACKAGE__) } = 1;

    confess sprintf(
        $_[0],
        defined $_[1] ? $_[1] : 'undef'
    );
}
## use critic

1;


lib/Net/Works/Address.pm
package Net::Works::Address;

use strict;
use warnings;

use namespace::autoclean;

our $VERSION = '0.23';

use Carp qw( confess );
use Math::Int128 0.06 qw( uint128 uint128_to_hex uint128_to_number );
use Net::Works::Types qw( PackedBinary Str );
use Net::Works::Util qw(
    _integer_address_to_binary
    _integer_address_to_string
    _string_address_to_integer
    _validate_ip_string
);
use Scalar::Util qw( blessed );
use Socket 1.99 qw( AF_INET AF_INET6 inet_pton inet_ntop );

use integer;

use overload (
    q{""} => '_overloaded_as_string',
    '<=>' => '_compare_overload',
    'cmp' => '_compare_overload',
);

use Moo;

with 'Net::Works::Role::IP';

has _binary => (
    is      => 'ro',
    reader  => 'as_binary',
    isa     => PackedBinary,
    lazy    => 1,
    builder => '_build_binary',
);

has _string => (
    is      => 'ro',
    reader  => 'as_string',
    isa     => Str,
    lazy    => 1,
    builder => '_build_string',
);

sub BUILD {
    my $self = shift;

    $self->_validate_ip_integer();

    return;
}

sub new_from_string {
    my $class = shift;
    my %p     = @_;

    my $str     = delete $p{string};
    my $version = delete $p{version};

    if ( defined $str && inet_pton( AF_INET, $str ) ) {
        $version ||= 4;
        $str = '::' . $str if $version == 6;
    }
    else {
        $version ||= 6;
        _validate_ip_string( $str, $version );
    }

    return $class->new(
        _integer => _string_address_to_integer( $str, $version ),
        version  => $version,
        %p,
    );
}

sub new_from_integer {
    my $class = shift;
    my %p     = @_;

    my $int     = delete $p{integer};
    my $version = delete $p{version};
    $version ||= ref $int ? 6 : 4;

    return $class->new(
        _integer => $int,
        version  => $version,
        %p,
    );
}

sub _build_string {
    my $self = shift;

    return _integer_address_to_string( $self->_integer() );
}

sub _build_binary { _integer_address_to_binary( $_[0]->as_integer() ) }

sub as_integer { $_[0]->_integer() }

sub as_ipv4_string {
    my $self = shift;

    return $self->as_string() if $self->version() == 4;

    confess
        'Cannot represent IP address larger than 2**32-1 as an IPv4 string'
        if $self->as_integer() >= 2**32;

    return ( ref $self )->new_from_integer(
        integer => $self->as_integer(),
        version => 4,
    )->as_string();
}

sub as_bit_string {
    my $self = shift;

    if ( $self->version == 6 ) {
        my $hex = uint128_to_hex( $self->as_integer() );
        my @ha  = $hex =~ /.{8}/g;
        return join q{}, map { sprintf( '%032b', hex($_) ) } @ha;
    }
    else {
        return sprintf( '%032b', $self->as_integer() );
    }
}

sub prefix_length { $_[0]->bits() }

sub mask_length { $_[0]->prefix_length() }

sub next_ip {
    my $self = shift;

    confess "$self is the last address in its range"
        if $self->as_integer() == $self->_max;

    return ( ref $self )->new_from_integer(
        integer => $self->as_integer() + 1,
        version => $self->version(),
    );
}

sub previous_ip {
    my $self = shift;

    confess "$self is the first address in its range"
        if $self->as_integer() == 0;

    return ( ref $self )->new_from_integer(
        integer => $self->as_integer() - 1,
        version => $self->version(),
    );
}

sub _compare_overload {
    my $self  = shift;
    my $other = shift;
    my $flip  = shift() ? -1 : 1;

    confess 'Cannot compare unless both objects are '
        . __PACKAGE__
        . ' objects'
        unless blessed $self
        && blessed $other
        && eval { $self->isa(__PACKAGE__) && $other->isa(__PACKAGE__) };

    return $flip * ( $self->as_integer() <=> $other->as_integer() );
}

__PACKAGE__->meta()->make_immutable();

1;

# ABSTRACT: An object representing a single IP (4 or 6) address

__END__

=head1 SYNOPSIS

  use Net::Works::Address;

  my $ip = Net::Works::Address->new_from_string( string => '192.0.2.1' );
  print $ip->as_string();     # 192.0.2.1
  print $ip->as_integer();    # 3221225985
  print $ip->as_binary();     # 4-byte packed form of the address
  print $ip->as_bit_string(); # 11000000000000000000001000000001
  print $ip->version();       # 4
  print $ip->prefix_length();   # 32

  my $next = $ip->next_ip();     # 192.0.2.2
  my $prev = $ip->previous_ip(); # 192.0.2.0

  if ( $next > $ip ) { print $ip->as_string(); }

  my @sorted = sort $next, $prev, $ip;

  my $ipv6 = Net::Works::Address->new_from_string( string => '2001:db8::1234' );
  print $ipv6->as_integer(); # 42540766411282592856903984951653831220

  my $ip_from_int = Net::Works::Address->new_from_integer(
      integer => "42540766411282592856903984951653831220"
  );

=head1 DESCRIPTION

Objects of this class represent a single IP address. It can handle both IPv4
and IPv6 addresses. It provides various methods for getting information about
the address, and also overloads the objects so that addresses can be compared
as integers.

For IPv6, it uses 128-bit integers (via Math::Int128) to represent the
numeric value of an address.

=head1 METHODS

This class provides the following methods:

=head2 Net::Works::Address->new_from_string( ... )

This method takes a C<string> parameter and an optional C<version>
parameter. The C<string> parameter should be a string representation of an IP
address.

The C<version> parameter should be either C<4> or C<6>, but you don't really need
this unless you're trying to force a dotted quad to be interpreted as an IPv6
address or to a force an IPv6 address colon-separated hex number to be
interpreted as an IPv4 address.

=head2 Net::Works::Address->new_from_integer( ... )

This method takes a C<integer> parameter and an optional C<version>
parameter. The C<integer> parameter should be an integer representation of an
IP address.

The C<version> parameter should be either C<4> or C<6>. Unlike with strings,
you'll need to set the version explicitly to get an IPv6 address.

=head2 $ip->as_string()

Returns a string representation of the address in the same format as
inet_ntop, e.g., "192.0.2.1", "::192.0.2.1", or "2001:db8::1234".

=head2 $ip->as_integer()

Returns the address as an integer. For IPv6 addresses, this is returned as a
L<Math::Int128> object, regardless of the value.

=head2 $ip->as_binary()

Returns the packed binary form of the address (4 or 16 bytes).

=head2 $ip->as_bit_string()

Returns the address as a string of 1's and 0's, like
"00000000000000000000000000010000".

=head2 $ip->as_ipv4_string()

This returns a dotted quad representation of an address, even if it's an IPv6
address. However, this will die if the address is greater than the max value
of an IPv4 address (2**32 - 1). It's primarily useful for debugging.

=head2 $ip->version()

Returns a 4 or 6 to indicate whether this is an IPv4 or IPv6 address.

=head2 $ip->prefix_length()

Returns the prefix length for the IP address, which is either 32 (IPv4) or 128
(IPv6).

=head2 $ip->bits()

An alias for C<< $ip->prefix_length() >>. This helps make addresses & network
objects interchangeable in some cases.

=head2 $ip->next_ip()

Returns the numerically next IP, regardless of whether or not it's in the same
subnet as the current IP.

This will throw an error if the current IP address it the last address in its
IP range.

=head2 $ip->previous_ip()

Returns the numerically previous IP, regardless of whether or not it's in the
same subnet as the current IP.

This will throw an error if the current IP address it the first address in
its IP range (address 0).

=head1 OVERLOADING

This class overloads comparison, allowing you to compare two objects and to
sort them (either as numbers or strings).

It also overloads stringification to call the C<< $ip->as_string() >> method.


lib/Net/Works/Role/IP.pm
package Net::Works::Role::IP;

use strict;
use warnings;
use namespace::autoclean 0.16;

our $VERSION = '0.23';

use Carp qw( confess );
use Math::Int128 qw( string_to_uint128 uint128 uint128_to_number );
use Net::Works::Types qw( Int IPInt IPVersion );
use Socket qw( AF_INET AF_INET6 );

use Moo::Role;

use integer;

has version => (
    is       => 'ro',
    isa      => IPVersion,
    required => 1,
);

has _integer => (
    is       => 'rw',
    writer   => '_set_integer',
    isa      => IPInt,
    required => 1,
);

has address_family => (
    is      => 'ro',
    isa     => Int,
    lazy    => 1,
    default => sub { $_[0]->version() == 6 ? AF_INET6 : AF_INET },
);

## no critic (ProhibitUnusedPrivateSubroutines)
{
    my %max = (
        4 => 0xFFFFFFFF,
        6 => string_to_uint128( '0x' . ( 'F' x 32 ) ),
    );

    sub _max {
        my $self    = shift;
        my $version = shift // $self->version();

        return $max{$version};
    }
}

sub bits { $_[0]->version() == 6 ? 128 : 32 }

sub _validate_ip_integer {
    my $self = shift;

    my $int = $self->_integer();

    # We don't need to check if it's too big with v6 because uint128 does not
    # allow a number larger than 2**128-1.
    if ( $self->version() == 6 ) {
        $self->_set_integer( uint128($int) )
            unless ref $int;
    }
    else {
        confess("$int is not a valid integer for an IP address")
            if $int >= 2**32;
        if ( ref $int ) {
            $self->_set_integer( uint128_to_number($int) );
        }
    }

    return;
}

# overload passes extra arguments to this sub for some reason
sub _overloaded_as_string {
    return $_[0]->as_string();
}

1;


lib/Net/Works/Types/Internal.pm
package Net::Works::Types::Internal;

use strict;
use warnings;
use namespace::autoclean;

use MooseX::Types -declare => [
    qw(
        IPInt
        IPVersion
        MaskLength
        PackedBinary
        UInt128
        )
];

use MooseX::Types::Moose qw( Int Str );

class_type UInt128, { class => 'Math::UInt128' };

subtype PackedBinary,
    as Str;

subtype IPInt,
    as Int | UInt128,
    where { $_ >= 0 },
    inline_as { $_[0]->parent()->_inline_check( $_[1] ) . " && ($_[1] >= 0)" },
    message { ( defined $_ ? $_ : 'undef' ) . ' is not a valid IP integer' };

subtype IPVersion,
    as Int;

subtype MaskLength,
    as Int,
    where { $_ >= 0 && $_ <= 128 },
    inline_as {
        $_[0]->parent()->_inline_check( $_[1] )
            . " && ($_[1] >= 0 && $_[1] <= 128)";
    },
    message {
        ( defined $_ ? $_ : 'undef' ) . ' is not a valid IP mask length';
    };

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Sane APIs for IP addresses and networks",
  "AUTHOR" => "Dave Rolsky <autarch\@urth.org>, Greg Oschwald <oschwald\@cpan.org>, Olaf Alders <oalders\@wundercounter.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Net-Works",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Net::Works",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Exporter" => 0,
    "List::AllUtils" => 0,
    "Math::Int128" => "0.06",
    "Moo" => 0,
    "Moo::Role" => 0,
    "Scalar::Util" => 0,
    "Socket" => "1.99",
    "Sub::Quote" => 0,
    "integer" => 0,
    "namespace::autoclean" => "0.16",
    "namespace::clean" => 0,
    "overload" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "B" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Math::BigInt" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.23",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "B" => 0,
  "Carp" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "List::AllUtils" => 0,
  "Math::BigInt" => 0,
  "Math::Int128" => "0.06",
  "Moo" => 0,
  "Moo::Role" => 0,
  "Scalar::Util" => 0,
  "Socket" => "1.99",
  "Sub::Quote" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.96",
  "integer" => 0,
  "namespace::autoclean" => "0.16",
  "namespace::clean" => 0,
  "overload" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
